docs/development/coding-style-guide.md

```diff
@@ -0,0 +1,246 @@
+
+Error strings are typically *chained*:
+
+```
+Deploying cluster: reading component config: parsing HCL: unexpected character '%'
+```
+
+Remember this fact when constructing error strings and think about how the final string might look
+to the user.
+
+#### Wrapping errors
+
+When "wrapping" an error in another error to provide context, the `%w` formatting verb should be
+used following the [Go 1.13 recommendations](https://blog.golang.org/go1.13-errors) to error
+handling:
+
+`fmt.Errorf("parsing config: %w", err)`
+
+#### Error string text
+
+Don't include words which describe failure in error strings. This includes words such as "failed",
+"error", "could not" etc. It is clear that an error indicates a failure. Instead, provide just the
+*context*, i.e. **what we were doing** when the error occurred (or the erroneous condition if the
+error is the last one in the "chain").
+
+Good:
+
+- `fmt.Errorf("parsing config: %w", err)`
+- `fmt.Errorf("deploying cluster: %w", err)`
+- `fmt.Errorf("reversing time: %w", err)`
+- `fmt.Errorf("space probe is offline")`
+
+Bad:
+
+- `fmt.Errorf("parsing config failed: %w", err)`
+- `fmt.Errorf("error encountered while deploying cluster: %w", err)`
+- `fmt.Errorf("could not reverse time: %w", err)`
+
+#### Capitalizing error strings
+
+Don't capitalize error strings when *returning* an error (remember - errors are chained).
+Capitalize the **beginning** of an error string when **logging** an error.
+
+Good:
+
+- `fmt.Errorf("parsing config: %w", err)`
+- `log.Errorf("Deploying cluster: %v", err)`
+
+Bad:
+
+- `fmt.Errorf("Parsing config: %w", err)`
+- `log.Errorf("deploying cluster: %v", err)`
+
+### Comments
+
+Follow the standard Go [guidelines](https://blog.golang.org/godoc) for comments in Go code. *In
+addition*, follow the guidelines below.
+
+#### Whitespace
+
+Include a space between `//` and the beginning of the comment.
+
+Good:
+
+```go
+// This is a comment.
+```
+
+Bad:
+
+```go
+//This is a comment.
+```
+
+#### Capitalization
+
+Capitalize the first letter of a Go comment unless the official Go guidelines require you to do
+otherwise.
+
+Good:
+
+```go
+// A very important constant.
+const this = "that"
+```
+
+Bad:
+
+```go
+// a very important constant.
+const this = "that"
+```
+
+#### Periods
+
+As a general rule, end all comments with a period. Possible exceptions are *very* short comments
+(up to 2-3 words which don't form a self-contained sentence) which are typically **inline**.
+
+Good:
+
+```go
+// Foo does this and also that.
+func Foo() {
+    ...
+    n := 3 // Default
+}
+```
+
+Bad:
+
+```go
+// Foo does this and also that
+func Foo() {
+    ...
+}
+```
+
+#### "nolint" comments
+
+"nolint" comments are a special type of comment used to tell the linter to ignore a specific
+linting rule. They are formatted differently from "regular" Go comments.
+
+Format "nolint" comments according to the following example:
+
+```go
+//nolint:funlen,gosec
+func foo() {
```

Comment:
As mentioned in #883, do we prefer nolint on top of the function, rather than inline with function, like `func foo() { //nolint:funlen` ?

Reply:
I find the inline version a bit strange, but no strong opinion here.

Reply:
https://golangci-lint.run/usage/false-positives/ suggests on top, so I suggest we should point here that having it inline is a bad example:
```
0 ✓ (134ms) 16:40:28 invidian@dellxps15mateusz ~/repos/kinvolk/lokomotive (invidian/more-cleanup)$ git grep -A1 nolint | grep func  | wc -l
26
0 ✓ (78.8ms) 16:40:30 invidian@dellxps15mateusz ~/repos/kinvolk/lokomotive (invidian/more-cleanup)$ git grep nolint | grep func | wc -l
7
```

I also think how does it play with function comment, like:
```go
// Foo does bar.
//
//nolint:funlen
func Foo() {
```

vs

```go
// Foo does bar
func Foo() { //nolint:funlen
```

I _think_ first case is still slightly better (less packed, so more readable).

Reply:
I also prefer the first example.

Reply:
How is it now (just pushed)?

Reply:
Looks great, thanks!